k3s-io/k3s
# Add support for dual-stack Pod/Service CIDRs and node IP addresses

#### Proposed Changes ####

* Add support for dual-stack cluster/service CIDRs and node addresses
* Begin transitioning the ip/externalip/hostname handoff (used by the K3s cloud provider) from labels to annotations, as labels do not allow the character-set necessary to pass through IPv6 addresses.
    * If a node has labels they will be maintained, but new nodes will use only annotations. The cloud provider will continue to check both, with annotations taking precedence.
* Dual-stack node addresses are not autodetected; if desired a dual-stack address list must be passed to --node-ip (and --node-external-ip as necessary). Kubeadm is taking the same approach: https://github.com/kubernetes/kubeadm/issues/1612#issuecomment-786540060

This opens the door to dual-stack operation, even if flannel and the kube-router NPC don't support it yet.

Note: Cluster-DNS and the Kubernetes apiserver `--advertise-address` themselves do not support dual-stack operation yet: https://github.com/kubernetes/kubeadm/issues/1612#issuecomment-772583989

#### Types of Changes ####

Feature

#### Verification ####

* Confirm that K3s works as usual when specifying only IPv4 CIDRs, or when using defaults
* Confirm that specifying dual-stack CIDRs raises an error if not run with `--disable-network-policy --flannel-backend=none`

Example test:
* Add `"ipv6": true, "fixed-cidr-v6": "fd7c:53a5:aef5:0::/64"` to /etc/docker/daemon.json and restart docker
* Run K3s server with IPv6:
    ```
    docker run --name=k3s-server-1 --rm -it --privileged rancher/k3s:BUILD_TAG --debug server --cluster-cidr=10.42.0.0/16,fd42::/48 --service-cidr=10.43.0.0/16,fd43::/112 --disable-network-policy --flannel-backend=none --node-ip=172.17.0.2,fd7c:53a5:aef5::242:ac11:2
    ```
* Check for dual-stack CIDRs and node addresses:
    ````
    docker exec -it k3s-server-1 kubectl get nodes -o yaml | grep -EA6 'spec:|addresses:'
      spec:
        podCIDR: 10.42.0.0/24
        podCIDRs:
        - 10.42.0.0/24
        - fd42::/64
        providerID: k3s://7a78f12c7fb1
        taints:
    --
          addresses:
        - address: 172.17.0.2
          type: InternalIP
        - address: fd7c:53a5:aef5::242:ac11:2
          type: InternalIP
        - address: 7a78f12c7fb1
          type: Hostname
    ````

#### Linked Issues ####

For #3158

#### Further Comments ####

While the CLI has been directly changed to use StringSlices instead of strings, I have opted to add separate fields to the  the Config types in order to retain backwards/forwards compatibility between builds with and without this change.
